LogicLayer/Friends/FriendsService.cs

```diff
@@ -0,0 +1,133 @@
+                .Include(u => u.SentRequests)
+                .SelectMany(u => u.SentRequests)
+                .Where(f => f.FriendshipStatus == FriendshipStatus.Lasts)
+                .Include(f => f.RequestedTo)
+                .Select(f => new UserDto(f.RequestedTo.Id, f.RequestedTo.UserName, f.RequestedTo.Email));
+
+            var friendsThatInvitedMe = user
+                .Include(u => u.ReceivedRequests)
+                .SelectMany(u => u.ReceivedRequests)
+                .Where(f => f.FriendshipStatus == FriendshipStatus.Lasts)
+                .Include(f => f.RequestedBy)
+                .Select(f => new UserDto(f.RequestedBy.Id, f.RequestedBy.UserName, f.RequestedBy.Email));
+
+            var allFriends = await friendsInvitedByMe // can I avoid using await here?
+                .Union(friendsThatInvitedMe)
+                .OrderBy(u => u.UserName) // maybe order by NormalizedUserName? However, I cannot do that with UserDto.
+                .ToListAsync();
+
+            return allFriends;
+        }
+
+        public async Task SendFriendRequestAsync(CreateFriendRequestDto friendRequest)
+        {
+            if (friendRequest.UserNameTo == friendRequest.UserNameFrom)
+            {
+                throw new ApplicationException("You cannot invite yourself.");
+            }
+
+            var requestedById = await _repository.GetUserIdByUsername(friendRequest.UserNameFrom);
+            var requestedToId = await _repository.GetUserIdByUsername(friendRequest.UserNameTo);
+
+            _repository.Friendships.Add(
+                new Friendship()
+                {
+                    RequestedById = requestedById,
+                    RequestedToId = requestedToId,
+                    FriendshipStatus = FriendshipStatus.PendingFriendRequest
+                });
+
+            try
+            {
+                await _repository.SaveChangesAsync();
+            }
+            catch (DbUpdateException e) when (e.InnerException is SqlException sqlException)
+            {
+                switch (sqlException.Number)
+                {
+                    case 50000:
+                        throw new FriendRequestAlreadyPendingException(
+                            "You have already been invited by this user. Check your inbox for the friend request.");
+                    case 50001:
+                        throw new FriendRequestAlreadyPendingException(
+                            "You have already sent this user a friend request.");
+                    case 50002:
+                        throw new FriendRequestAlreadyFinalizedException("You are already friends.");
+                    default:
+                        throw new ArgumentOutOfRangeException();
+                }
+            }
+
+            //send email with link
+        }
+
+        public Task<FriendRequestDto> GetFriendRequestAsync(string friendRequestId)
+        {
+            var id = int.Parse(friendRequestId);
```

Comment:
When I first wrote the interface I expected IDs to actually be strings. They're not, so there's no reason for the interface to accept strings and then parse them - change the parameter type to `int`.

Reply:
Did it and also changed your file to be compatible with this.